contracts/modules/SwapHub.sol

```diff
@@ -0,0 +1,168 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+pragma solidity ^0.8.0;
+
+import "../BaseLogic.sol";
+import "../swapHandlers/ISwapHandler.sol";
+
+/*
+
+SwapHub is a generic swapping interface where users can select their desired swapping handler
+without any changes needed to Euler contracts.
+
+When a user invokes a swap, the input amount (or maximum input) is transferred to the handler
+contract. The handler contract should then perform the swap by whatever means it chooses,
+then transfer back any remaining input and all the output. SwapHub will ensure that the
+amounts returned satisfy the user's slippage settings and process the corresponding
+withdrawal and deposit on behalf of the user.
+
+*/
+
+contract SwapHub is BaseLogic {
+    struct SwapCache {
+        address accountIn;
+        address accountOut;
+        address eTokenIn;
+        address eTokenOut;
+        AssetCache assetCacheIn;
+        AssetCache assetCacheOut;
+        uint preBalanceIn;
+        uint preBalanceOut;
+    }
+
+    constructor(bytes32 moduleGitCommit_) BaseLogic(MODULEID__SWAPHUB, moduleGitCommit_) {}
+
+    function swap(uint subAccountIdIn, uint subAccountIdOut, address swapHandler, ISwapHandler.SwapParams memory params) external nonReentrant {
+        SwapCache memory cache = initSwap(subAccountIdIn, subAccountIdOut, params);
+
+        emit RequestSwapHub(
+            cache.accountIn,
+            cache.accountOut,
+            params.underlyingIn,
+            params.underlyingOut,
+            params.amountIn,
+            params.amountOut,
+            params.mode,
+            swapHandler
+        );
+
+        uint amountOut = swapInternal(cache, swapHandler, params);
+
+        // Process deposit
+        uint amountOutDecoded = decodeExternalAmount(cache.assetCacheOut, amountOut);
+        uint amountOutInternal = underlyingAmountToBalance(cache.assetCacheOut, amountOutDecoded);
+        cache.assetCacheOut.poolSize = decodeExternalAmount(cache.assetCacheOut, cache.preBalanceOut + amountOut);
+        AssetStorage storage assetStorageOut = eTokenLookup[cache.eTokenOut];
+        increaseBalance(assetStorageOut, cache.assetCacheOut, cache.eTokenOut, cache.accountOut, amountOutInternal);
+        logAssetStatus(cache.assetCacheOut);
+
+        // Check liquidity
+        checkLiquidity(cache.accountIn);
+        if (assetStorageOut.users[cache.accountOut].owed != 0) checkLiquidity(cache.accountOut);
```

Comment:
how about adding the condition for `accountIn != accountOut`? otherwise, liquidity might be checked twice for the same account

Reply:
nice catch, thanks